examples/nlp/language_modeling/megatron_change_num_partitions.py

```diff
@@ -12,36 +12,204 @@
+        elif 'key_value.weight' in param_name and megatron_legacy:
+            split_dim = partitions[0][idx].data.shape[0]
+            if split_dim % (tp_size * 2) != 0:
+                raise ValueError(
+                    f"Can not split K,V parameter {param_name} with shape {param.shape} into tensor parallel size {tp_size}. Not divisible by {tp_size * 2}."
+                )
+            tp_qkv_splits = torch.chunk(partitions[0][idx].data, tp_size * 2, dim=0)
+            split = []
+            for i in range(tp_size):
+                tp_qkv = torch.cat([tp_qkv_splits[item] for item in range(i, tp_size * 2, tp_size)])
+                split.append(tp_qkv)
+        # Regular split for Megatron and NeMo-Megatron models.
+        else:
+            split = torch.split(partitions[0][idx].data, param.shape[0], dim=0)
+
+    return split
+
+
+def merge_partition(model, partitions: Dict[int, List[List[torch.Tensor]]], write_path: str = None):
+    # Extract the pp_rank and number of modules per tp rank in each pp rank
+    pp_ranks = list(partitions.keys())
+    pp_lens = []
+    for pp_rank in pp_ranks:
+        partition_pp = partitions[pp_rank]
+        max_len = max([len(x) for x in partition_pp])  # Perform max as we need to iterate through all modules
+        pp_lens.append(max_len)
+
+    total_params_merged = len([p for p in model.parameters()])
+    pp_total_len = sum(pp_lens)
+    logging.info(f"Total layers in Merged Model: {total_params_merged}")
+
+    og_pp_split_rank = 0
+    if pp_total_len > total_params_merged:
+        og_pp_split_rank = model.cfg.get('pipeline_model_parallel_split_rank', 0)
+
     idx = 0
+    pp_rank = 0
+    global_idx = 0
+
+    # During merge - model is TP 1 PP 1 model with all parameters present in correct order.
+    # Merge the parameters of the various PP X TP Y models into the TP 1 PP 1 model.
     for name, param in model.named_parameters():
-        if param.shape == partitions[0][idx].shape:
-            concated = partitions[0][idx].data
-        elif param.shape[0] == partitions[0][idx].shape[0]:
-            concated = torch.cat([partitions[i][idx].data for i in range(len(partitions))], dim=-1)
+        # Since the PP ranks each contain the list of all their TP rank parameters
+        # We need to detect if we need to move to the next PP rank when we run out of tensors in current PP rank
+        # Reset the index so that it indexes the new pp rank tensor list correctly
+        if idx >= pp_lens[pp_rank]:
+            pp_rank += 1
+            idx = 0
+
+            # For EncDec models, after the encoder-decoder PP split occurs,
+            # the vocab and positional embeddings are duplicated across the PP ranks at the
+            # beginning of the decoder rank. We can skip them during the merge step.
+            if pp_total_len > total_params_merged:
+                if og_pp_split_rank > 0 and og_pp_split_rank == pp_rank:
+                    logging.info(
```

Comment:
can we still check that the next 2 idx are vocab and positional embeddings? (some like if 'vocab' in name) Want to avoid cases where things can break silently in case new models classes get added that don't follow this regime

Reply:
There won't be silent errors, end of the config has exact param count check, every single param has to match or it will raise an error.

I can't rework the logic for this, because I cannot make assumptions of what the parameter types of names are. It would make pp change much more complicated

Reply:
sounds good